source/agora/consensus/Validation.d

```diff
@@ -217,3 +219,236 @@ unittest
     // Signature verification must be error
     assert(!tx2.isValid(findOutput), format("Transaction signature is not validated %s", tx2));
 }
+
+/*******************************************************************************
+
+    Check the validity of a block.
+
+    A block is considered valid if:
+        - its height is the previous block height + 1
+        - its prev_hash is the previous block header's hash
+        - the number of transactions in the block are equal to Block.TxsInBlock
+        - the merkle root in the header matches the re-built merkle tree root
+          based on the included transactions in the block
+        - all the the transactions pass validation, which implies:
+            - signatures are authentic
+            - the inputs spend an output which must be found with the
+              findOutput() delegate
+
+    Note that checking for transactions which double-spend is the responsibility
+    of the findOutput() delegate. During validation, whenever this delegate is
+    called it should also keep track of the used UTXOs, thereby marking
+    it as a spent output. See the `findNotDoubleSpent` function in the
+    unittest for an example.
+
+    Params:
+        block = the block to check
+        prev_height = the height of the previous block which this
+                      block should point to
+        findOutput = delegate to find the referenced unspent UTXOs with
+
+    Returns:
+        true if the block is considered valid
+
+*******************************************************************************/
+
+public bool isValid (const ref Block block, in ulong prev_height,
+    in Hash prev_hash, UtxoFinder findOutput) nothrow @safe
+{
+    import std.algorithm;
+
+    // special case for the genesis block
+    if (block.header.height == 0)
+    {
+        return () @trusted
+        {
+            try { return block == getGenesisBlock(); }
+            catch (Exception) { assert(0); }
+        }();
+    }
+
+    if (block.header.height != prev_height + 1)
+        return false;
+
+    if (block.header.prev_block != prev_hash)
+        return false;
+
+    if (block.txs.length != Block.TxsInBlock)
+        return false;
+
+    if (block.txs.any!(tx => !tx.isValid(findOutput)))
+        return false;
+
+    static Hash[] merkle_tree;
```

Comment:
I think I need to make this a parameter, to make it at least somewhat re-entrant. What do you think @Geod24?